### PR TITLE
Simplify/deabstract module_host_actor and wasmer_module

### PR DESCRIPTION
**Pull request by coolreader18**
# Description of Changes

# API and ABI

 - [ ] This is a breaking change to the module ABI
 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API

*If the API is breaking, please state below what will break*

# Expected complexity level and risk

*How complicated do you think these changes are? Grade on a scale from 1 to 5,
where 1 is a trivial change, and 5 is a deep-reaching and complex change.*

*This complexity rating applies not only to the complexity apparent in the diff,
but also to its interactions with existing and future code.*

*If you answered more than a 2, explain what is complex about the PR,
and what other components it interacts with in potentially concerning ways.*

### CODE REVIEW

**Comment by kim** [comment]
This will conflict with #267

**Comment by kulakowski** [comment]
Not to jump the gun, but I took a peek and think this is going to be helpful to people clicking around in and reading this code.

**Comment by kim** [comment]
Not saying this is bad, but it will make @mamcx’s life harder to get #267 over the line. Also as a heads up: with the changes in #267 much of the init/update code becomes obsolete (for one because we can no longer update index changes). Please let me know when this becomes ready, I think I’m mostly the one familiar with that code.

**Comment by coolreader18** [comment]
More than happy to wait on #267 to be merged first.

**Comment by coolreader18** [comment]
What's the status with #267 re:being split into multiple parts? I can't tell whether the chunks that are relevant to wasmer_module are in master already or not.

**Comment by coolreader18** [comment]
benchmarks please

**Comment by coolreader18** [comment]
lol. well

**Comment by coolreader18** [comment]
I guess that confirms that works @kurtismullins

**Comment by coolreader18** [comment]
benchmarks please

**Comment by coolreader18** [comment]
benchmarks please

**Comment by mamcx** [comment]
> What's the status with #267 re:being split into multiple parts? I can't tell whether the chunks that are relevant to wasmer_module are in master already or not.

Is still waiting for approving, and it will take some time because was split in several other PRs...

**Comment by kim** [comment]
Alright, then let’s merge this?

**Comment by kim** [comment]
@coolreader18 Maybe you could add this from #386: https://github.com/clockworklabs/SpacetimeDB/pull/386/files#diff-840af797a96dc965f1b75269b1c343356ecb6a0b4e28b84c611a0db514e9d3faR113-R127 ?

I think it is still better meanwhile to create multi-column indexes than to ignore them.

**Comment by coolreader18** [comment]
Maybe as a separate PR based on this one :grimacing: ? unless I missed that when I was rebasing and it's in master already. Just cause it's kinda its own feature rather than a refactor like this.

**Comment by coolreader18** [comment]
benchmarks please

**Comment by coolreader18** [comment]
cool it works :))))

File: crates/core/src/db/relational_db.rs
Comment (kulakowski) [inline crates/core/src/db/relational_db.rs:373]: I'd keep the `instrument` bit.

File: crates/core/src/db/datastore/traits.rs
Comment (kulakowski) [inline crates/core/src/db/datastore/traits.rs:296]: I don't love this code being here. But I don't think there's a great place to put it when we are in this state of defining a pile of concrete types, and traits (that aren't really used, relationaldb directly makes a Locking or whatever). Curious how you feel about it.

File: crates/core/src/db/update.rs
Comment (kulakowski) [inline crates/core/src/db/update.rs:23]: I didn't scrutinize this code, am trusting that it's functionally unchanged.

**Review by kulakowski — COMMENTED** [review]
Ok I think this is good and ready. It does a bit of unrelated cleanups but I like them all.

I'm not hitting approve just to make sure that we are all squared up on which PR lands before whichever other PR, but I'll hit it as as soon as someone acknowledges that the plan is ready.

File: crates/core/src/db/relational_db.rs
Comment (kulakowski) [reply crates/core/src/db/relational_db.rs:373]: This is the only thing to fix before landing, then, I think.

**Review by kulakowski — APPROVED** [review]

File: crates/core/src/db/relational_db.rs
Comment (coolreader18) [reply crates/core/src/db/relational_db.rs:373]: IMO it's really unnecessary - it's a utility function that we have no need to track and only serves to clutter up logs/graphs. If you feel strongly about it I can add it back, but I really see no value in it.

File: crates/core/src/db/datastore/traits.rs
Comment (coolreader18) [reply crates/core/src/db/datastore/traits.rs:296]: Yeah, I think there's better ways to go about this, but ideally that would come with a general reorganizing of all these types.

**Comment by coolreader18** [comment]
Merged. @joshua-spacetime if you'd like help fixing #486 to work with this I'd be happy to help

File: crates/core/src/db/datastore/traits.rs
Comment (cloutiertyler) [reply crates/core/src/db/datastore/traits.rs:296]: I also don't love this code being here FWIW

File: crates/core/src/db/relational_db.rs
Comment (kulakowski) [reply crates/core/src/db/relational_db.rs:373]: It's not that I do or don't see the value in it. It's that it is unrelated to the stated point of the PR.

**Pull request merged**
